### ISSUE TITLE
Add working directory property

### ISSUE THREAD

**Issue by rejeep** [flycheck/flycheck#312]
As discussed in https://github.com/flycheck/flycheck/issues/308, we should add a property called `:cwd` that will change the `default-directory` when running a checker. The property takes a single argument, which is a function that will be called **each** time the checker runs and it will return a string path.

I guess it makes to fallback and not change the default directory if the function returns `nil`.

Example:

```lisp
(flycheck-define-checker foo
  :cwd (lambda ()
         (f-traverse-upwards
          (lambda (path)
            (f-dir? (f-expand ".cask" path)))
          (f-parent (f-this-file)))))
```

Note that I have noticed that setting the `default-directory` for a sub-process did not always work unless you specified the absolute path with an ending slash. The function `f-full` does exactly this.

**Comment by rbraunstein** [comment]
+1

**Comment by s3bs** [comment]
+1, I'm using flycheck with a custom checker. I modified my flycheck sources such that it changes default-directory in the (let ... scope where the checker is executed. For me a static :cwd is sufficient at the moment. Also as discussed in #308 I added functionality to change the process-environment temporary to add environment variables needed by the checker, which are also static.

**Comment by lunaryorn** [comment]
@s3bs I'm sorry that you have to go so far as to modify our sources, but currently this feature has a rather low priority on our backlog.  To me, #774 and #800 are much more important.  They are also much larger and complex than this one, so it'll take a long time before I get to this issue.

I'd love to see a PR for this feature :relaxed:

**Comment by s3bs** [comment]
@lunaryorn I actually like to tinker once in a while with elisp code. If I'd know more elisp I'd give you a hand. I put my changes that work for me as a fork in my github repository in the cwd branch if that helps anyone in the meantime. I'm not quite sure what you mean with 'PR'?!

**Comment by rbraunstein** [comment]
PR is short for pull request.  Check github help or google. You are pretty
close.

**Comment by BernardNotarianni** [comment]
I did modify a checker to change directory using the `:predicate` property.
Any idea to improve this to make the working directory coming back to the original location?

```elisp
(with-eval-after-load "flycheck"
  (flycheck-define-checker elixir-mix
    "An Elixir syntax checker using the Elixir interpreter.
     See URL `http://elixir-lang.org/'."
    :command ("mix"
              "compile"
              source)
    :predicate is-mix-project-p
    :error-patterns
    ((error line-start "** (" (zero-or-more not-newline) ") "
            (zero-or-more not-newline) ":" line ": " (message) line-end)
     (warning line-start
              (one-or-more (not (syntax whitespace))) ":"
              line ": "
              (message)
              line-end))
    :modes elixir-mode)
  (add-to-list 'flycheck-checkers 'elixir-mix))

(defun is-mix-project-p ()
  (let ((mix-project-root (locate-dominating-file (buffer-file-name) "mix.exs")))
    (if mix-project-root (cd mix-project-root) nil)))

```

**Comment by lunaryorn** [comment]
@BernardNotarianni You may have luck with hooking into `flycheck-after-syntax-check-hook`, but you're really abusing `:predicate` here; we'll not support that.  As far as I'm concerned I'd strongly advise you to wait for this PR being merged.

 If you have further questions about this hack or need further help please let us discuss elsewhere (i.e. in the issue you have referenced).  Please let us keep this thread focused on the discussion of this pull request.

Thank you.

**Comment by jaseemabid** [comment]
@lunaryorn Any chance of adding a cwd? `rebar3 dialyzer` needs to be run at the root of the project as well.

**Comment by lunaryorn** [comment]
@jaseemabid We have an open PR at #813 which still needs work and testing afaik.

**Comment by jaseemabid** [comment]
@lunaryorn Ah. Thank you.

**Comment by jaseemabid** [comment]
@lunaryorn In the specific project I'm working on, all source resides in a single src folder and all I have to do is add `cd .. && <specific linter command>` into `~/bin/foo` and set `:cmd` to `foo`. 

Is there a flycheck way to avoid the external bin file and add `cd &&` into the `:cmd` somehow?

**Comment by lunaryorn** [comment]
@jaseemabid Not currently, no.  We do not even have `:cwd` property currently.  I'm sorry.

**Comment by lunaryorn** [comment]
Fixed as of #973
